File: common/src/main/java/com/ssblur/scriptor/word/action/SmiteAction.java
package com.ssblur.scriptor.word.action;

import com.ssblur.scriptor.api.word.Action;
import com.ssblur.scriptor.api.word.Descriptor;
import com.ssblur.scriptor.data_components.ScriptorDataComponents;
import com.ssblur.scriptor.helpers.ItemTargetableHelper;
import com.ssblur.scriptor.helpers.targetable.EntityTargetable;
import com.ssblur.scriptor.helpers.targetable.Targetable;
import com.ssblur.scriptor.word.descriptor.power.StrengthDescriptor;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.EntityType;
import net.minecraft.world.entity.LightningBolt;

public class SmiteAction extends Action {
  @Override
  public void apply(Targetable caster, Targetable targetable, Descriptor[] descriptors) {
    double strength = 1;
    for(var d: descriptors) {
      if(d instanceof StrengthDescriptor strengthDescriptor)
        strength += strengthDescriptor.strengthModifier();
    }

    var itemTarget = ItemTargetableHelper.getTargetItemStack(targetable);
    if(!itemTarget.isEmpty()) {
      itemTarget.set(ScriptorDataComponents.CHARGES, (int) Math.floor(strength));
      return;
    }

    ServerLevel level = (ServerLevel) targetable.getLevel();
    LightningBolt bolt = new LightningBolt(EntityType.LIGHTNING_BOLT, level);
    if(caster instanceof EntityTargetable entityTargetable && entityTargetable.getTargetEntity() instanceof ServerPlayer player)
      bolt.setCause(player);
    bolt.setPos(targetable.getTargetPos());
    level.addFreshEntity(bolt);
  }

  @Override
  public Cost cost() { return new Cost(12, COSTTYPE.ADDITIVE); }

}

File: common/src/main/kotlin/com/ssblur/scriptor/word/action/SmiteAction.kt
package com.ssblur.scriptor.word.action

import com.ssblur.scriptor.api.word.Action
import com.ssblur.scriptor.api.word.Descriptor
import com.ssblur.scriptor.data.components.ScriptorDataComponents
import com.ssblur.scriptor.helpers.ItemTargetableHelper
import com.ssblur.scriptor.helpers.targetable.EntityTargetable
import com.ssblur.scriptor.helpers.targetable.Targetable
import com.ssblur.scriptor.word.descriptor.power.StrengthDescriptor
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.EntityType
import net.minecraft.world.entity.LightningBolt
import kotlin.math.floor

class SmiteAction: Action() {
  override fun apply(caster: Targetable, targetable: Targetable, descriptors: Array<Descriptor>) {
    var strength = 1.0
    for (d in descriptors) {
      if (d is StrengthDescriptor) strength += d.strengthModifier()
    }

    val itemTarget = ItemTargetableHelper.getTargetItemStack(targetable)
    if (!itemTarget.isEmpty) {
      itemTarget.set(ScriptorDataComponents.CHARGES, floor(strength).toInt())
      return
    }

    val level = targetable.level as ServerLevel
    val bolt = LightningBolt(EntityType.LIGHTNING_BOLT, level)
    if (caster is EntityTargetable && caster.targetEntity is ServerPlayer) bolt.cause =
      (caster.targetEntity as ServerPlayer)
    bolt.setPos(targetable.targetPos)
    level.addFreshEntity(bolt)
  }

  override fun cost() = Cost(12.0, COSTTYPE.ADDITIVE)
}
